### PR TITLE
debaml: native structural class coercion + defaults (Mcoerce-d PR 2/3)

### PR DESCRIPTION
**Pull request by invakid404**
## Mcoerce-d PR 2/3 — structural class defaults

De-BAML epic #546, slice **Mcoerce-d PR 2 of 3**. This is the deep/riskiest slice: it ports BAML's `coerce_class.rs` (the non-array subset) into `coerceClass`. PR 1 (stringification + literal extraction) is already merged (#562); the union revisit is PR 3.

### What it does

`coerceClass` now reproduces BAML's class value **byte-for-byte** for the deterministic non-union subset, on top of the existing field-matching/first-wins/ExtraKey/nullable-clean-only scaffolding:

- **Object field assignment** — input-key order, first-field match (`matches_string_to_string`, no substring), keep-first duplicates, unmatched keys flagged `ExtraKey` (cost 1) and omitted, fields emitted in schema order under canonical names.
- **Single-field OBJECT implied-key** — when no key matches the lone field, the whole object is coerced into it (`ImpliedKey`, cost 2), no `ExtraKey`.
- **Single-field SCALAR/null inferred-object** — a non-object non-array value is absorbed into the lone field (`ImpliedKey` cost 2 + `InferedObject` cost 0).
- **Missing optional → null** (`OptionalDefaultFromNoValue`, cost 1) — omitted from bytes (InjectAbsentOptionals re-adds it) but score-flagged.
- **Required-field defaults** via `TypeIR::default_value` (`defaultValue`): list→`[]`, map→`{}`, primitive-null→`null`, first-defaultable-union-arm, tuple — `DefaultFromNoValue` (cost 100).
- **Present required-map non-object → `{}`** (`DefaultButHadUnparseableValue`, cost 2): `coerce_map` errors on a non-object (`error_unexpected_type`), a proven failure with a deterministic default.

Every new flag threads into `coerceFlags`, so the nullable-union clean-only rule stays conservative. A class child that now succeeds through these paths is **kept** in a list/map instead of declining the whole collection. A claimed child error (enum/literal substring tie) **propagates** as the class error (status parity).

### Boundaries held (verified)

- **ARRAY input to a class DECLINES** — `coerce_array_to_singular` / multi-candidate `pick_best` is M3. (This also fixes a latent over-claim: the old blanket `typeMismatch` was wrong for `[{...}]` inputs BAML coerces.)
- **Required non-defaultable missing field DECLINES** — BAML errors, but native falls back to stay strictly non-over-claiming.
- **No union-family broadening / no `pick_best`** (PR 3).
- **Fixtures 40 / 42 / 43 stay fallback** — they decline at the *schema gate* (`checkSupportedUnionShape` rejects single-field / mixed class unions), so `coerceClass` becoming lenient cannot flip them. A unit test (`TestCoerceClass_Fixture40ShapeStaysFallback`) pins this and confirms both single-field arms *do* absorb the scalar via inferred-object (exactly why the union must stay fallback).

### Validation

- `internal/debaml` unit tests: new `coerce_class_test.go` covers every structural path (assign / implied-key / inferred-object / optional-null / each default kind / non-defaultable-fails / claimed-error propagation / collection flips / nullable-clean-only / `defaultValue`), plus updated map/class tests.
- **11 live-captured** parse-recovery fixtures (125–135) — all diff-green native-vs-BAML (`88 claimed, 44 fallback`, full suite passes). Wants captured with `BAMLFUZZ_PARSE_CAPTURE=1`, never hand-written.
- Over-claim guard: 1200 random generated schemas run through the native-vs-BAML parse differential — 0 divergence.
- `gofmt` / `go vet ./...` / `go vet -tags=integration` / `go test ./internal/debaml ./bamlutils ./worker ./dynclient/...` / `GOWORK=off go test ./codegen` / `go run ./cmd/verify-framework-adapter` all green.

### Differential dispositions (every flip justified)

Flipped **fallback → claimed** (native now byte-matches BAML):

| fixture | input | BAML/native output | why |
|---|---|---|---|
| 125 `single_field_class_scalar_inferred` | `{"u":5}` | `{"u":{"value":5}}` | inferred-object |
| 126 `single_field_class_object_implied_key` | `{"u":{"foo":5}}` | `{"u":{"value":"{foo: 5}"}}` | implied-key + JsonToString |
| 127 `class_missing_optional_null` | `{"u":{"name":"Ada"}}` | `{"u":{"name":"Ada","nick":null}}` | OptionalDefaultFromNoValue |
| 128 `class_required_list_default_from_no_value` | `{"u":{}}` | `{"u":{"items":[]}}` | list DefaultFromNoValue |
| 129 `class_required_map_default_from_no_value` | `{"u":{}}` | `{"u":{"m":{}}}` | map DefaultFromNoValue |
| 130 `class_required_null_default_from_no_value` | `{"u":{}}` | `{"u":{"n":null}}` | null DefaultFromNoValue |
| 131 `class_map_field_default_but_had_unparseable_value` | `{"u":{"m":"bad"}}` | `{"u":{"m":{}}}` | DefaultButHadUnparseableValue |
| 132 `list_single_field_class_scalar_kept` | `{"u":[5,6]}` | `{"u":[{"value":5},{"value":6}]}` | element inferred-object kept |
| 133 `map_string_single_field_class_scalar_value_kept` | `{"u":{"a":5,"b":6}}` | `{"u":{"a":{"value":5},"b":{"value":6}}}` | value inferred-object kept |
| 134 `list_class_required_default_value_kept` | `{"u":[{}]}` | `{"u":[{"items":[]}]}` | element default kept |
| 135 `map_string_class_required_default_value_kept` | `{"u":{"k":{}}}` | `{"u":{"k":{"items":[]}}}` | value default kept |

Also flipped in unit tests: `TestCoerceMap_NonObjectFieldDefaults` / `TestParse_MapNonObjectFieldDefaults` (`{"scores":"x"|5|[1,2,3]}` → `{"scores":{}}`), single-field-class inferred/implied cases, and top-level/array-to-class now **declines** (was a claimed error — the old claim was buggy for `[{...}]`).

Stayed **fallback** (unchanged): 40, 42, 43, 39, 53 (union scoring, M3), and the required-missing-field / array-to-class declines.

### Note: `class_required_union_first_defaultable_arm` omitted

The union-first-defaultable-arm default (`default_value`'s Union branch) is faithfully implemented and unit-tested (`TestDefaultValue`), but is **not reachable as a claimed fixture**: a *required* non-nullable union must be an M2c safe family (literal/class union — no defaultable arm), and a *nullable* union field is optional, so an absent one takes the `OptionalDefaultFromNoValue` null path, not `default_value`. So no supported schema yields a non-null union default via a class field.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/invakid404/baml-rest/pull/563?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Expanded parsing recovery for single-field classes (including inferred values from scalar/object shapes) and stronger missing-field defaulting for required/optional fields.
  * Improved structural default handling for lists, maps, and null values when inputs are incomplete or malformed.
* **Bug Fixes**
  * Non-object values for required map fields now recover by producing empty/default map results instead of failing.
  * Updated fallback vs decline behavior for top-level array inputs to be more consistent.
* **Tests**
  * Added new parse-recovery fuzz fixtures and extended unit/integration coverage for these edge cases.

### CODE REVIEW

File: internal/debaml/coerce.go [hunk 1]
```diff
@@ -1223,11 +1446,16 @@ func provenListItemError(b *schema.Bundle, elem schema.Type, item value) bool {
...
 		// A multi-field class whose fields are ALL required flat leaves has no
-		// default_value for any field, so a genuine SCALAR input leaves every
-		// required field unfilled → BAML error_missing_required_field. An ARRAY
-		// element defers to coerce_array_to_singular (M3); a single-field class, or
-		// a class with any defaultable (list/map/optional) field, defers to
-		// Mcoerce-d — so those are NOT proven.
+		// default_value for any field and is not single-field, so a genuine SCALAR
+		// input can be neither inferred-object-absorbed nor default-filled — every
+		// required field stays unfilled → BAML error_missing_required_field (a
+		// PROVEN skip, which coerceClass itself now claims via missingRequiredField).
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce.go]: Addressed in `42b400a2` (duplicate of the CodeRabbit finding on the same comment). The stale comment(s) that claimed `coerceClass` "claims via missingRequiredField" (a non-existent symbol) are reworded: `coerceClass` **declines** via `unsupported()` on a missing non-defaultable required field (the local `missingRequired` flag), and the proven list/map skip is established independently by `provenListItemError`/`classAllRequiredFlatLeaf` — exactly your suggested wording.

File: internal/debaml/coerce_class_test.go [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+	}{
+		{"list", listT, "[]"},
+		{"map", mapT, "{}"},
+		{"null", nullT, "null"},
+		// Union: FIRST defaultable arm (list) wins in declaration order.
+		{"union-list-first", schema.Type{Kind: schema.TypeUnion, Union: &schema.UnionType{Variants: []schema.Type{listT, strT}}}, "[]"},
+		// Union: a non-defaultable arm is skipped; the map arm defaults.
+		{"union-skip-nondefaultable", schema.Type{Kind: schema.TypeUnion, Union: &schema.UnionType{Variants: []schema.Type{strT, mapT}}}, "{}"},
+		// Nullable union with no defaultable non-null arm -> the appended null arm.
+		{"union-nullable-null", schema.Type{Kind: schema.TypeUnion, Union: &schema.UnionType{Variants: []schema.Type{intT}, Nullable: true}}, "null"},
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce_class_test.go:288]: Addressed in `42b400a2` (duplicate of the CodeRabbit tuple-coverage finding). `TestDefaultValue` now has `schema.TypeTuple` cases: tuple-single `[null]`, tuple-all-defaultable `[[],{}]` (recursive branch + item joining), and tuple-has-nondefaultable (early-return nil/false) — matching your suggested cases.

File: internal/debaml/coerce_class_test.go [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+	// Duplicate matching key: first value wins.
+	mustParse(t, s, `{"a":1,"a":2,"b":"x"}`, `{"a":1,"b":"x"}`)
+
+	// ExtraKey is score-bearing: a class with an extra key is not clean.
+	b, ct := classBundle(t, s, "Baml_Rest_DynamicOutput")
+	cf := &coerceFlags{}
+	if _, err := coerceClass(b, ct.Name, ct.Mode, objVal(fld("a", numV("1")), fld("b", strVv("x")), fld("extra", numV("9"))), cf); err != nil {
+		t.Fatalf("coerceClass extra: %v", err)
+	}
+	if !cf.isFlagged() {
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce_class_test.go:68]: Confirmed addressed in `42b400a2` (as you auto-detected). The inferred-object null path is now covered: intBox/strBox with bare null → declines (coerce_int/coerce_string on null is error_unexpected_null → fall back), and an optional lone field absorbs null → `{value:null}` (ImpliedKey).

File: internal/debaml/coerce_class_test.go [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+		t.Fatalf("inferred-object: %v", err)
+	}
+	if !cf.isFlagged() {
+		t.Errorf("inferred-object must flag cf (ImpliedKey, score-bearing)")
+	}
+
+	// A scalar that the lone field cannot coerce -> DECLINE (native cannot prove
+	// BAML's outcome). "hello" -> int is a proven BAML error, but native declines.
+	requireUnsupported(t, intBox, `{"u":"hello"}`)
+}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce_class_test.go:153]: Addressed in `42b400a2`: `TestCoerceClass_SingleFieldScalarInferredObject` now covers a bare JSON null through the inferred-object path (intBox/strBox null → declines via error_unexpected_null fall-back; optional lone field absorbs null → `{value:null}`).

File: internal/debaml/coerce_class_test.go [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+		{"literal", litT},
+		// A non-nullable union of non-defaultable arms (int | string).
+		{"union-all-nondefaultable", schema.Type{Kind: schema.TypeUnion, Union: &schema.UnionType{Variants: []schema.Type{intT, strT}}}},
+	}
+	for _, c := range nonDefaultable {
+		if _, ok := defaultValue(c.t); ok {
+			t.Errorf("defaultValue(%s) must be non-defaultable (ok=false)", c.name)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce_class_test.go:321]: Addressed in `42b400a2`: `TestDefaultValue` now covers `schema.TypeTuple` — tuple-single `[null]`, tuple-all-defaultable `[[],{}]` (recursive branch + item joining), and tuple-has-nondefaultable (early-return nil/false).

File: internal/debaml/coerce.go [hunk 6]
```diff
@@ -1223,11 +1446,16 @@ func provenListItemError(b *schema.Bundle, elem schema.Type, item value) bool {
...
+		// Everything else is NOT proven: a single-field class (Mcoerce-d inferred-
+		// object / implied-key) or a class with any defaultable (list/map/optional/
+		// defaultable-union) field now SUCCEEDS through coerceClass and is KEPT
+		// before reaching here (or DECLINES as indeterminate → whole-list fallback);
+		// a valObject may coerce or default; an ARRAY element defers to
+		// coerce_array_to_singular (pick_best, M3).
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce.go:1460]: Addressed in `42b400a2`: the stale comment (non-existent `missingRequiredField`, wrongly saying "claimed") is reworded to reflect that `coerceClass` **declines** via `unsupported()` on a missing non-defaultable required field (local `missingRequired` flag), with the proven collection skip established independently by `provenListItemError`/`classAllRequiredFlatLeaf`. Comment-only, no behavior change.
